### PR TITLE
prov/rxm: Implement support for credit based flow control

### PR DESCRIPTION
**Pull request by ooststep**
To mitigate some performance issues caused by packet loss due to CQ overrun, this set of changes implements an optional flow control mechanism in RxM. Utility providers that want to take advantage of this flow control should implement the new ofi_ops_flow_ctrl structure and return this struct from ops_open.  Currently, only verbs implements this interface and all other utility providers will not user credit based flow control by default.

### CODE REVIEW

File: prov/rxm/src/rxm_ep.c [hunk 1]
```diff
@@ -1577,6 +1587,32 @@ void rxm_ep_progress_deferred_queue(struct rxm_ep *rxm_ep,
...
+				if (ret == -FI_EAGAIN)
+					break;
+				rxm_cq_write_error(def_tx_entry->rxm_ep->util_ep.rx_cq,
+						   def_tx_entry->rxm_ep->util_ep.rx_cntr,
+						   def_tx_entry->rndv_read.rx_buf->
+							recv_entry->context, ret);
+				break;
+			}
+			rxm_ep_dequeue_deferred_tx_queue(def_tx_entry);
+			ofi_buf_free(def_tx_entry->credit_msg.tx_buf);
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c:1613]: We need to wait for the completion before freeing the buffers for reuse.

File: prov/verbs/src/verbs_ep.c [hunk 2]
```diff
@@ -37,15 +37,32 @@
...
+void vrb_add_credits(struct fid_ep *ep_fid, size_t credits)
+{
+	struct vrb_ep *ep;
+	struct util_cq *cq;
+
+	ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+	cq = ep->util_ep.rx_cq;
+
+	cq->cq_fastlock_acquire(&cq->cq_lock);
+	ep->peer_rq_credits += credits;
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c:49]: The comment says that peer_rq_credits is protected by the send cq lock, but we use the receive cq lock here.  I think the send cq lock is correct.

File: prov/verbs/src/verbs_ep.c [hunk 3]
```diff
@@ -60,7 +77,21 @@ ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
...
+
+	if (++ep->rq_credits_avail >= domain->threshold) {
+		credits_to_give = ep->rq_credits_avail;
+		ep->rq_credits_avail = 0;
+	} else {
+		credits_to_give = 0;
+	}
 	cq->util_cq.cq_fastlock_release(&cq->util_cq.cq_lock);
+
+	if (credits_to_give && domain->send_credits(&ep->util_ep.ep_fid, credits_to_give)) {
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c:89]: check line length

File: prov/verbs/src/verbs_ep.c [hunk 4]
```diff
@@ -113,12 +156,24 @@ ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr)
...
+	ep->sq_credits++;
 freebuf:
 	ofi_buf_free(ctx);
+	if (ep->rq_credits_avail >= domain->threshold) {
+		credits_to_give = ep->rq_credits_avail;
+		ep->rq_credits_avail = 0;
+	}
 unlock:
 	cq->util_cq.cq_fastlock_release(&cq->util_cq.cq_lock);
+	if (credits_to_give && domain->send_credits(&ep->util_ep.ep_fid, credits_to_give)) {
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c:172]: check line length

File: prov/verbs/src/verbs_ep.c [hunk 5]
```diff
@@ -70,35 +101,47 @@ ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
...
 			vrb_save_wc(cq, &wc);
 
-		if (!cq->credits || !ep->tx_credits)
+		if (!cq->credits || !ep->sq_credits || !ep->peer_rq_credits) {
 			goto freebuf;
+		}
+	}
+
+	if ((wr->opcode == IBV_WR_SEND || wr->opcode == IBV_WR_SEND_WITH_IMM ||
+	     wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM) &&
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c:132]: A macro or static inline would make this a easier to read -- e.g. vrb_wr_consumes_recv()

**Comment by ooststep** [comment]
I believe i have addressed all review comments so far.

File: prov/rxm/src/rxm_conn.c [hunk 6]
```diff
@@ -786,22 +786,109 @@ int rxm_cmap_alloc(struct rxm_ep *rxm_ep, struct rxm_cmap_attr *attr)
...
+	struct rxm_conn *rxm_conn =
+		container_of(ep->fid.context, struct rxm_conn, handle);
+	struct rxm_ep *rxm_ep = rxm_conn->handle.cmap->ep;
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	struct rxm_tx_base_buf *tx_buf;
+	struct iovec iov;
+	struct fi_msg msg;
+	ssize_t ret;
+
+	tx_buf = (struct rxm_tx_base_buf *) rxm_tx_buf_alloc(
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c:800]: Every place that rxm_tx_buf_alloc() is called is preceded by a cast to some other structure.  This tells me that rxm_tx_buf_alloc() should really just return a void *.  This should be a separate change, and I can add that after merging this.

File: prov/rxm/src/rxm_conn.c [hunk 7]
```diff
@@ -786,22 +786,109 @@ int rxm_cmap_alloc(struct rxm_ep *rxm_ep, struct rxm_cmap_attr *attr)
...
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"unable to allocate TX entry for deferred CREDIT mxg\n");
+			ofi_buf_free(tx_buf);
+			return -FI_ENOMEM;
+		}
+
+		def_tx_entry->credit_msg.tx_buf = tx_buf;
+		rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
+		return FI_SUCCESS;
+	}
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c:839]: nit: I would move the defer code outside of the if statement as:
```
    if (!ret)
        return FI_SUCCESS;

defer:
    def_tx_entry = ...
```
It 'undents' most of the code and avoids having a goto jump inside an if statement.

File: prov/rxm/src/rxm_conn.c [hunk 8]
```diff
@@ -786,22 +786,110 @@ int rxm_cmap_alloc(struct rxm_ep *rxm_ep, struct rxm_cmap_attr *attr)
...
+	if (!ret && flow_ctrl_ops) {
+		rxm_ep->flow_ctrl_ops = flow_ctrl_ops;
+		rxm_ep->flow_ctrl_ops->set_threshold(rxm_domain->msg_domain,
+						     rxm_ep->msg_info->rx_attr->size / 2);
+		rxm_ep->flow_ctrl_ops->set_send_handler(rxm_domain->msg_domain,
+							rxm_send_credits);
+	} else if (ret == -FI_ENOSYS)
+		rxm_ep->flow_ctrl_ops = &rxm_no_ops_flow_ctrl;
+	else
+		goto err;
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_conn.c:889]: I suggest we do:
```
if (!ret) {
    assert(flow_ctrl_ops);
    ...
} else {
    rxm_ep->flow_ctrl_ops = ...
}
```

File: prov/rxm/src/rxm_cq.c [hunk 9]
```diff
@@ -1406,11 +1426,12 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
 		}
 	} while ((ret > 0) && (++comp_read < rxm_ep->comp_per_progress));
 
-	if (OFI_UNLIKELY(!dlist_empty(&rxm_ep->deferred_tx_conn_queue))) {
+	if (!dlist_empty(&rxm_ep->deferred_tx_conn_queue)) {
 		dlist_foreach_container_safe(&rxm_ep->deferred_tx_conn_queue,
 					     struct rxm_conn, rxm_conn,
 					     deferred_conn_entry, conn_entry_tmp)
-			rxm_ep_progress_deferred_queue(rxm_ep, rxm_conn);
+			if (rxm_conn->handle.state == RXM_CMAP_CONNECTED)
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: You added the check to the progress function, but kept it here.

**Comment by ooststep** [comment]
@shefty noticed one of my fixes didn't get added, should be good now.

**Comment by shefty** [comment]
I think my review overlapped with one of the changes that you were pushing at the time.

**Comment by ooststep** [comment]
> I think my review overlapped with one of the changes that you were pushing at the time.

Yes, it was the comment about the check to the progress function that I did not completely push the fix for. it's updated now and should address your comment.  I'll work on a follow on PR for the `rxm_tx_buf_alloc` change.  Finishing smoke tests of the defer change and will push that update shortly.

**Pull request merged**